CONTRIBUTING.md (#59)
mcb-dev/mCodingBot: CONTRIBUTING.md

# Contributing

## Tooling Guide

### Poetry

> mCodingBot bot uses poetry for package managing. Poetry can be installed [with thier official instructions](https://python-poetry.org/docs/#installation).

Reviewer: ```suggestion
mCodingBot bot uses poetry for package managing. Poetry can be installed [with their official instructions](https://python-poetry.org/docs/#installation).
```

> Before you start coding you have to activate your env with `poetry shell`.

Reviewer: ```suggestion
Before you start coding, you have to activate your env with `poetry shell`.
```

The bot can be run with `python -m mcodingbot`.

### Linting

> these commands can be run to lint your code.

Reviewer: ```suggestion
These commands can be run to lint your code:
```

```
# check if your code is properly linted
nox -s lint

# check for typing issues
nox -s mypy

# fix linting errors with black, codespell, and isort.
nox -s apply-lint
```

Your code must pass the `mypy` and `lint` nox pipelines to be merged.

## Style Guide

### Imports
All modules use `import mod`.
The typing module is the only exception where you are allowed to do `from typing import T`.

> Code from the bot itself itself is always imported as `from bot.a import b`.

Reviewer: ```suggestion
Code from the bot itsel is always imported as `from bot.a import b`.
```

# Contributing Guide

## Initial Setup

To begin contributing to mCodingBot, you must first fork the repository. After forking, clone:

```
git clone <url to your fork>
cd mCodingBot
```

Next, you need to set up poetry.

```
curl -sSL https://install.python-poetry.org | python3 -
poetry install
```

See the [installation guide](https://python-poetry.org/docs/#installation) for alternative methods.

## Running the Bot

> First, if you're making changes that require the database, you will need to set up Postgres. Once installed, create a user named `mcodingbot`, and then a database called `mcodingbot` that is owned by that user.

Reviewer 2: Isn't it `PostgreSQL`...?
```suggestion
First, if you're making changes that require the database, you will need to set up PostgreSQL. Once installed, create a user named `mcodingbot`, and then a database called `mcodingbot` that is owned by that user.
```

Reviewer 2: > Both names are in use :v
> 
> From [Wikipedia](https://en.wikipedia.org/wiki/PostgreSQL):
> 
> > _PostgreSQL, also known as Postgres, [...]._

oops, my bad

Author: I prefer postgres here. It's shorter.

Reviewer 2: > I prefer postgres here. It's shorter.

lol it's shorter by two chars though...

Author: It's also shorter by two syllables

> To run the bot, you first need to set up your `config.json` file. Look at [config.json.example](https://github.com/mcb-dev/mCodingBot/blob/main/config.json.example) for an example configuration. Make sure to set `no_db_mode: true` if you chose not to set up Postgres. If you did set up Postgres, you only need to put your password (if any) in the config.json.

Reviewer 2: ```suggestion
To run the bot, you first need to set up your `config.json` file. Look at [config.json.example](https://github.com/mcb-dev/mCodingBot/blob/main/config.json.example) for an example configuration. Make sure to set `no_db_mode: true` if you chose not to set up Postgres. If you did set up Postgres, you only need to put your password (if any) in the `config.json`.
```
I guess?

Author: +1

> Once the config is setup, you can run the bot with `poetry run python3 -m mcodingbot`.

Reviewer 2: Out of scope, but wouldn't it be simpler to add the script in the `[tools.poetry.scripts]` section of the `pyproject.toml` file?
Then it could be run by simply doing `poetry run mcodingbot`
In that case this could be the subject of another PR, but I'd be interested in your thoughts on this.

Reviewer 2: Created a new issue (73) so you can comment there

Author: Most definitely out of scope so im resolving this

> Once the config is setup, you can run the bot with `poetry run python3 -m mcodingbot`.

Reviewer: ```suggestion
Once the config is set up, you can run the bot with `poetry run python3 -m mcodingbot`.
```

## Contributing

> To create a pull request to mCodingBall, first create a new branch for your changes.

Reviewer 2: ```suggestion
To create a pull request to mCodingBot, first create a new branch for your changes.
```
Well played lol but I read code attentively

> To create a pull request to mCodingBall, first create a new branch for your changes.

Reviewer: ```suggestion
To create a pull request to mCodingBall, first, create a new branch for your changes.
```

Author: This feels wrong to me because it makes "first" look like a parenthetical. Imo the best thing to do here is reword the entire sentence if you aren't happy with it.

```
git checkout -b <new-branch-name>
```

> Then, open the source in your editor of choice, and make changes. After making changes, make sure that they pass the pipelines by running `nox`.

Author: I don't think "open the source" makes sense. I think "repository", "project", or "source code" are better options.

```
poetry run nox  # run all pipelines
poetry run nox -s mypy  # check only the mypy pipeline
poetry run nox -s apply-lint  # run black and isort

> poetry run nos -s lint  # checks if flake8, black, and isort all pass

Author: I find this harder to read. The short lines make you look really far over to see what they do.

> poetry run nos -s lint  # checks if flake8, black, and isort all pass

Reviewer: > I find this harder to read. The short lines make you look really far over to see what they do.

idk for me the current one is more cluttered and harder to read

alternative methods.

## Running the Bot

First, if you're making changes that require the database, you will need to set
up Postgres. Once installed, create a user named `mcodingbot`, and then a
database called `mcodingbot` that is owned by that user.

To run the bot, you first need to set up your `config.json` file. Look at
[config.json.example](https://github.com/mcb-dev/mCodingBot/blob/main/config.json.example)
for an example configuration. Make sure to set `no_db_mode: true` if you chose not to
set up Postgres. If you did set up Postgres, you only need to put your password
(if any) in the `config.json`.

Once the config is set up, you can run the bot with
`poetry run python3 -m mcodingbot`.

## Contributing

To create a pull request to mCodingBot, first create a new branch for your
changes.

```
git checkout -b <new-branch-name>
```

Then, open the repository in your editor of choice, and make changes. After
making changes, make sure that they pass the pipelines by running `nox`.

```bash
poetry run nox                # run all pipelines
poetry run nox -s mypy        # check only the mypy pipeline
poetry run nox -s apply-lint  # run black and isort

> poetry run nos -s lint        # checks if flake8, black, and isort all pass

Reviewer 3: ```suggestion
poetry run nox -s lint        # checks if flake8, black, and isort all pass
```
